src/armor_layers.cpp

```diff
@@ -112,387 +106,499 @@ item_penalties get_item_penalties( std::list<item>::const_iterator worn_item_it,
-    }
-
-    if( worn_item.has_flag( "HOOD" ) ) {
-        description_stack.push_back( _( "It has a hood." ) );
-    }
-    if( worn_item.has_flag( "POCKETS" ) ) {
-        description_stack.push_back( _( "It has pockets." ) );
-    }
-    if( worn_item.has_flag( "WATERPROOF" ) ) {
-        description_stack.push_back( _( "It is waterproof." ) );
-    }
-    if( worn_item.has_flag( "WATER_FRIENDLY" ) ) {
-        description_stack.push_back( _( "It is water friendly." ) );
-    }
-    if( worn_item.has_flag( "FANCY" ) ) {
-        description_stack.push_back( _( "It looks fancy." ) );
-    }
-    if( worn_item.has_flag( "SUPER_FANCY" ) ) {
-        description_stack.push_back( _( "It looks really fancy." ) );
-    }
-    if( worn_item.has_flag( "FLOTATION" ) ) {
-        description_stack.push_back( _( "You will not drown today." ) );
-    }
-    if( worn_item.has_flag( "OVERSIZE" ) ) {
-        description_stack.push_back( _( "It is very bulky." ) );
-    }
-    if( worn_item.has_flag( "SWIM_GOGGLES" ) ) {
-        description_stack.push_back( _( "It helps you to see clearly underwater." ) );
-    }
-    if( worn_item.has_flag( "SEMITANGIBLE" ) ) {
-        description_stack.push_back( _( "It can occupy the same space as other things." ) );
-    }
+    protected:
+        item_penalties penalties;
+        int encumber;
+        std::string name;
+        std::list<item>::iterator itm_iter;
+
+    public:
+        layering_item_info( std::list<item>::iterator itm_i, player &p, body_part bp ) {
+            penalties = get_item_penalties( itm_i, p, bp );
+            encumber = itm_i->get_encumber( p );
+            name = itm_i->tname();
+            itm_iter = itm_i;
+        }
 
-    return description_stack;
-}
+        std::list<item>::iterator get_iter() {
+            return itm_iter;
+        }
 
-} //namespace
+        bool operator ==( const layering_item_info &o ) const {
+            return itm_iter == o.itm_iter;
+        }
 
-struct layering_item_info {
-    item_penalties penalties;
-    int encumber;
-    std::string name;
-
-    // Operator overload required to leverage vector equality operator.
-    bool operator ==( const layering_item_info &o ) const {
-        // This is used to merge e.g. both arms into one entry when their items
-        // are equivalent.  For that purpose we don't care about the exact
-        // penalities because they will list different body parts; we just
-        // check that the badness is the same (which is all that matters for
-        // rendering the right-hand list).
-        return this->penalties.badness() == o.penalties.badness() &&
-               this->encumber == o.encumber &&
-               this->name == o.name;
-    }
+        int get_col_value( sort_armor_columns col ) {
+            item itm = *itm_iter;
+            int ret_val;
+            switch( col ) {
+                case column_warmth:
+                    ret_val = itm.get_warmth();
```

Comment:
```suggestion
                    return itm.get_warmth();
```
Same for all the other placements below. That way you can cut the size of this function in half (get rid of all the `break` statements).